lms/static/scripts/frontend_apps/services/test/client-rpc-test.js

```diff
@@ -0,0 +1,117 @@
+import { $imports, ClientRpc } from '../client-rpc';
+
+describe('ClientRpc', () => {
+  let clientConfig;
+  let FakeServer;
+  let fakeSidebarWindow;
+  let fakeRpcCall;
+  let fakeServerInstance;
+
+  beforeEach(() => {
+    clientConfig = { showHighlights: true };
+
+    fakeSidebarWindow = {
+      frame: {},
+      origin: 'https://client.hypothes.is',
+    };
+
+    fakeServerInstance = {
+      register: sinon.stub(),
+      sidebarWindow: fakeSidebarWindow,
+    };
+
+    fakeRpcCall = sinon.stub();
+
+    FakeServer = sinon.stub().returns(fakeServerInstance);
+
+    $imports.$mock({
+      '../../postmessage_json_rpc': {
+        Server: FakeServer,
+        call: fakeRpcCall,
+      },
+    });
+  });
+
+  afterEach(() => {
+    $imports.$restore();
+  });
+
+  function createClientRpc() {
+    return new ClientRpc({
+      allowedOrigins: ['https://client.hypothes.is'],
+      clientConfig,
+    });
+  }
+
+  it('initializes RPC server that accepts requests from specified origins', () => {
+    createClientRpc();
+    assert.calledWith(FakeServer, ['https://client.hypothes.is']);
+  });
+
+  it('registers "requestConfig" RPC handler that returns client config', () => {
+    createClientRpc();
+    assert.calledWith(fakeServerInstance.register, 'requestConfig');
+    const callback = fakeServerInstance.register.args[0][1];
+    assert.equal(callback(), clientConfig);
+  });
+
+  it('registers "requestGroups" RPC handler', () => {
+    createClientRpc();
+    assert.calledWith(fakeServerInstance.register, 'requestGroups');
+  });
+
+  describe('setGroups', () => {
+    it('sets the groups returned by "requestGroups" RPC handler', async () => {
+      const clientRpc = createClientRpc();
+      clientRpc.setGroups(['groupA', 'groupB']);
+
+      const callback = fakeServerInstance.register.args[1][1];
+      const groups = await callback();
```

Comment:
This could perhaps use a clarifying comment (I can infer what's being tested here but an explicit comment would be helpful).

Reply:
I've restructured this code a little to hopefully make it clearer and also avoid an assumption about the order in which RPC methods are registered.